### PR TITLE
Work stealing scheduler

### PR DESCRIPTION
**Pull request by arthurp**
Add a custom work stealing scheduler and make it the default.

Also make a number of other changes used in developing and debugging the scheduler and bugs found in other code while working on the scheduler.

I will run full tests in the next few days. But I figured you could glance at the code before I finish.

### CODE REVIEW

**Comment by arthurp** [comment]
I have done what I think are the last needed fixes for things exposed by the new scheduler or by the longer benchmarks.

Please review.

File: OrcScala/src/orc/run/core/Token.scala [hunk 1]
```diff
@@ -849,11 +922,18 @@ private class LongCounter(private var value: Long) {
...
+    override def initialValue() = false
+  }
+  */
+
   private val currentTokenDebugId = new ThreadLocal[LongCounter]() {
     override def initialValue() = new LongCounter(0L)
   }
   def getNextTokenDebugId(runtime: OrcRuntime): Long =
-    /* FIXME:This adverse coupling to runtime should be removed */
+    /* FIXME:This adverse coupling to runtime should be removed. Why not add a runtimeThreadId to the Orc trait? */
```
Comment (jthywiss) [inline OrcScala/src/orc/run/core/Token.scala]: It seemed a bit low-level for the Orc trait, but I may do that in the end.

File: OrcScala/src/orc/run/StandardOrcRuntime.scala [hunk 2]
```diff
@@ -19,8 +19,9 @@ import orc.run.extensions._
 
 class StandardOrcRuntime(engineInstanceName: String) extends Orc(engineInstanceName)
   with StandardInvocationBehavior
-  //with OrcWithWorkStealingScheduler
-  with OrcWithThreadPoolScheduler
+  with OrcWithWorkStealingScheduler
```
Comment (jthywiss) [inline OrcScala/src/orc/run/StandardOrcRuntime.scala]: We're stable enough for this?

File: OrcScala/.settings/org.scala-ide.sdt.core.prefs [hunk 3]
```diff
@@ -58,7 +58,7 @@ g=vars
 nameHashing=false
 no-specialization=false
 nowarn=false
-optimise=false
+optimise=true
```
Comment (jthywiss) [inline OrcScala/.settings/org.scala-ide.sdt.core.prefs]: oh!

File: OrcScala/src/orc/run/StandardOrcRuntime.scala (hunk 2 above)
Comment (arthurp) [inline OrcScala/src/orc/run/StandardOrcRuntime.scala]: The only way to know is to test it. I'm not sure it's stable enough for release. But it's stable enough for master.

File: OrcScala/src/orc/run/core/Token.scala (hunk 1 above)
Comment (arthurp) [inline OrcScala/src/orc/run/core/Token.scala]: I don't agree. I think logging utilities are very reasonable to be there. Orc represents access to an Orc implementation not some abstract idea of an Orc runtime. If you wanted we could have an abstract interface that is outward facing and one with details like this that is used throughout the implementation tree.

File: OrcTests/test_data/performance/threads.orc [hunk 4]
```diff
@@ -2,7 +2,7 @@
 This program creates 2^20 (or about 1 million) threads
 and waits for them to terminate.
 -}
-val N = 13
+val N = 18
```
Comment (arthurp) [inline OrcTests/test_data/performance/threads.orc]: No. 2^20 takes to long on the old thread scheduler to be practical for benchmarking. I think we can play with it later and change it again.

File: OrcScala/src/orc/run/extensions/WorkStealingScheduler.scala [hunk 5]
```diff
@@ -1,9 +1,7 @@
 //
-// WorkStealingScheduler.scala -- Scala trait OrcWithWorkStealingScheduler and class OrcWorkStealingExecutor
+// WorkStealingScheduler.scala -- A work-stealing scheduler for Orc
 // Project OrcScala
 //
-// Created by amp on Nov 4, 2015.
```
Comment (arthurp) [inline OrcScala/src/orc/run/extensions/WorkStealingScheduler.scala]: Nah, I created it. I just didn't know the date and think these comments are less than useless. I'll make up some random numbers for it.

**Pull request merged**